### ISSUE TITLE
new rule: disable-operator

### ISSUE THREAD

**Issue by dimitropoulos** [eslint/eslint#8216]
**Please describe what the rule should do:**
disallow an operator.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**
The only operator I have a use-case for needing to disable is `in`

I often review code that does:
```js
if ('b' in a) {
  return a.b.c;
}
```

where the intent is actually:
```js
if (a && a.b) {
  return a.b.c;
}
```

Where `a = { b: undefined }`, the first one will throw, and the second will not.

Of course, you have the other use of the `in` operator: loops.  In this case, I think this rule could be well served to have a different option for disabling that case individually.  Although it hadn't occurred to me at first, if there was a rule to warn when using `for in` constructions, I would probably use it now that `for of` exists (as I can in almost all cases disregard `for in`). 

examples that would be in a test include
```js
var a = { b: undefined };
if ('b' in a) {
  return a.b.c;
}

if (a && 'b' in a) {
  return a.b.c;
}

const x = 'b' in a ? true : false;

// with arrays too
var d = [0, 1, 2];
if (0 in d) {
  return d[0];
}

const z = 0 in d ? true : false;
```

**Why should this rule be included in ESLint (instead of a plugin)?**
This rule is for a feature of the language 'proper', and therefore does not belong in a plugin.

**Comment by not-an-aardvark** [comment]
Thanks for the rule proposal. For what it's worth, this will be possible to do with `no-restricted-syntax` when [selector support](https://github.com/eslint/eslint/pull/7833) is added. I'm wondering if it would be better to wait until that's possible so that we don't have too many redundant rules.

**Comment by platinumazure** [comment]
@dimitropoulos For your use case in particular, this would be pretty easy to write as a custom rule. You would just need to listen for `BinaryExpression` nodes and report a lint violation if the token is `in`. Maybe that would solve your issue while allowing us to deliberate on the best way to handle operator reporting?

```js
module.exports = {
    meta: {
        // Metadata goes here
    },

    create: function(context) {
        return {
            BinaryExpression: function(node) {
                if (node.operator === "in") {
                    context.report({
                        node,
                        message: "Do not use 'in' operator."
                    });
                }
            }
        }
    }
};
```

**Comment by vitorbal** [comment]
I agree with @not-an-aardvark that once we get selection support for no-restricted-syntax, these kinds of checks would be easy to configure.
Meanwhile, the solution proposed by @platinumazure seems pretty straightforward. Would that be enough to solve your use-case, @dimitropoulos?

**Comment by dimitropoulos** [comment]
sweet! thanks guys! I'll try this out, and if I have any other trouble, I'll reopen this.  Didn't know about the selectors work!  looks super super cool.
